src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+                                               max_gene_hits_per_multiple_gene_alignments: int,
+                                               primary_alignment_record_tags_set: Set[AlignmentRecordTags]) -> \
+            List[List[AlignmentRecordTags]]:
+
+        primary_alignment_record_tags_list = list(primary_alignment_record_tags_set)
+
+        multiple_gene_alignment_tags_list: List[List[AlignmentRecordTags]] = list()
+        for _ in range(num_multiple_gene_alignments):
+            random_primary_alignment = self.rng.choice(primary_alignment_record_tags_list)
+            random_primary_cell_barcode: str = random_primary_alignment.cell_barcode
+            novel_molecule_barcode: str = self._generate_unique_random_alignment_tag(
+                primary_alignment_record_tags_set,
+                gene_name=random_primary_alignment.gene_name,
+                cell_barcode=random_primary_cell_barcode).molecule_barcode
+            num_gene_hits = self.rng.randint(low=2, high=max_gene_hits_per_multiple_gene_alignments + 1)
+            gene_name_hits = self.rng.choice(self.to_be_used_gene_names, replace=False, size=num_gene_hits)
+            multiple_gene_alignment_tags_list.append(
+                [AlignmentRecordTags(random_primary_cell_barcode, novel_molecule_barcode, gene_name)
+                 for gene_name in gene_name_hits])
+        return multiple_gene_alignment_tags_list
+
+    @staticmethod
+    def _load_gene_names(annotation_file: str) -> List[str]:
+        return [k for k, v in sorted(gtf.extract_gene_names(annotation_file).items(), key=operator.itemgetter(1))]
+
+    def _generate_random_cell_barcode(self, length: int = 16):
+        return self._generate_random_genomic_sequences(length)
+
+    def _generate_random_molecule_barcode(self, length: int = 10):
+        return self._generate_random_genomic_sequences(length)
+
+    def _generate_random_genomic_sequences(self, length: int):
+        return ''.join(self.rng.choice(['A', 'C', 'T', 'G'], size=length))
+
+
+def _get_sorted_count_matrix(count_matrix: np.ndarray, row_index: np.ndarray, col_index: np.ndarray)\
+        -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    sorted_row_indices = [idx for idx, _ in sorted(enumerate(row_index), key=operator.itemgetter(1))]
+    sorted_col_indices = [idx for idx, _ in sorted(enumerate(col_index), key=operator.itemgetter(1))]
+    return (count_matrix[sorted_row_indices, :][:, sorted_col_indices],
+            row_index[sorted_row_indices],
+            col_index[sorted_col_indices])
+
+
+@pytest.mark.parametrize('query_sort_order', ['CB_UB_GE_n', 'n'])
+def test_count_matrix_from_bam(query_sort_order):
+    # instantiate a test data generator
+    synthetic_data_generator = SyntheticBarcodedBAMGenerator(
+        _test_num_cells,
+        _test_max_genes,
+        _test_annotation_file,
+        _test_gene_expression_rate)
+
+    with tempfile.TemporaryDirectory() as _test_temp_dir:
```

Comment:
I'd add a `finally` clause to explicitly `shutil.rmtree` the temp dir -- on some operating systems, these files get left around until restart, so it's nice to clean up after ourselves. :)

Reply:
I did it just to be safe, but I have a feeling that this is redundant. There are several instances of improper usage of `tempfile.TemporaryDirectory()` in `sctools` and that is likely to be the cause of improper cleanup.

Reply:
Cool, made #45.